src/matrix.c

```diff
@@ -0,0 +1,250 @@
+    for (R_len_t jj = startcol; jj <= endcol; jj++) {
+      memcpy(pmat + vecIdx, pcol[jj], sizeof(double)*n);
+      vecIdx += n;
+    }
+  }
+  
+  UNPROTECT(1);
+  return mat;
+}
+
+SEXP asmatrix_complex(SEXP dt, R_xlen_t matlen, R_xlen_t n) {
+  // Create output matrix, allocate memory, and create a pointer to it.
+  SEXP mat = PROTECT(allocVector(CPLXSXP, matlen));
+  Rcomplex *pmat; 
+  pmat = COMPLEX(mat);
+  
+  // Create an array of pointers for the individual columns in DT 
+  R_len_t dtncol = length(dt);
+  Rcomplex **pcol;
+  pcol = (Rcomplex **) R_alloc(dtncol, sizeof(Rcomplex *));
+  for (R_len_t jj = 0; jj < dtncol; jj++) {
+    pcol[jj] = COMPLEX(VECTOR_ELT(dt, jj));
+  }
+  
+  // Set up parallel OMP chunk
+  R_len_t chunksize = dtncol/getDTthreads() + 1; 
+  #pragma omp parallel num_threads(getDTthreads())
+  { 
+    // Determine which columns in DT we're iterating through on this thread
+    int threadnum = omp_get_thread_num();
+    R_len_t startcol = threadnum * chunksize;
+    R_len_t endcol = startcol + chunksize - 1;
+    if (endcol >= dtncol) 
+      endcol = dtncol - 1;
+    
+    // Determine where to fill in the matrix vector
+    R_xlen_t vecIdx = startcol * n;
+    
+    // Iterate through columns in DT, copying the contents to the matrix column
+    for (R_len_t jj = startcol; jj <= endcol; jj++) {
+      memcpy(pmat + vecIdx, pcol[jj], sizeof(Rcomplex)*n);
+      vecIdx += n;
+    }
+  }
+
+UNPROTECT(1);
+return mat;
+}
+
+SEXP asmatrix_character(SEXP dt, R_xlen_t matlen, R_len_t n) {
+  // Create output matrix, allocate memory, and create a pointer to it
+  SEXP mat = PROTECT(allocVector(STRSXP, matlen)); // output matrix
+  SEXP pcol; // pointers to casted R objects
+  
+  // Determine number of columns and rows
+  R_len_t dtncol = length(dt);
+
+  // Iterate through dt and copy into mat 
+  R_xlen_t vecIdx = 0; // counter to track place in vector underlying matrix
+  for (R_len_t jj = 0; jj < dtncol; jj++) {
+    pcol = VECTOR_ELT(dt, jj);
+    for (R_len_t ii = 0; ii < n; ii++) {
+      SET_STRING_ELT(mat, vecIdx, STRING_ELT(pcol, ii));
+      vecIdx++;
+    }
+  }
+  
+  UNPROTECT(1);
+  return mat;
+}
+
+SEXP asmatrix_list(SEXP dt, R_xlen_t matlen, R_len_t n) {
+  // Create output matrix, allocate memory, and create a pointer to it
+  SEXP mat = PROTECT(allocVector(VECSXP, matlen)); // output matrix
+  SEXP pcol; // pointers to casted R objects
+  
+  // Determine number of columns and rows
+  R_len_t dtncol = length(dt);
+
+  // Iterate through dt and copy into mat 
+  R_xlen_t vecIdx = 0; // counter to track place in vector underlying matrix
+  for (R_len_t jj = 0; jj < dtncol; jj++) {
+    pcol = VECTOR_ELT(dt, jj);
+    for (R_len_t ii = 0; ii < n; ii++) {
+      SET_VECTOR_ELT(mat, vecIdx, VECTOR_ELT(pcol, ii));
+      vecIdx++;
+    }
+  }
+  
+  UNPROTECT(1);
+  return mat;
+}
+
+// Dispatch function for different atomic types
+SEXP asmatrix(SEXP dt) {
```

Comment:
this fun is perfect place for a verbose timing, an example in https://github.com/Rdatatable/data.table/blob/dd6f306d7791d6ae3b10bd582906720f6645badc/src/fwhich.c#L267-L295

Reply:
Should I add a verbose argument to as.matrix, or just rely on `options(datatable.verbose)`?

I could put three timings in, one for preprocessing, one for the C code, and a total time. E.g. the messages would look something like:

"as.matrix: time taken to check and coerce column types: %.3f s"
"as.matrix: time taken to copy data into matrix: %.3f s"
"as.matrix: total time taken: %.3f s"

Reply:
I could not see any instances of system.time() being used in any of the R/ files so I have only implemented the verbose timing as suggested at the C level